### PR TITLE
[SPARK-16062][SPARK-15989][SQL] Fix two bugs of Python-only UDTs

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

There are two related bugs of Python-only UDTs. Because the test case of second one needs the first fix too. I put them into one PR. If it is not appropriate, please let me know.
### First bug: When MapObjects works on Python-only UDTs

`RowEncoder` will use `PythonUserDefinedType.sqlType` for its deserializer expression. If the sql type is `ArrayType`, we will have `MapObjects` working on it. But `MapObjects` doesn't consider `PythonUserDefinedType` as its input data type. It causes error like:

```
import pyspark.sql.group
from pyspark.sql.tests import PythonOnlyPoint, PythonOnlyUDT
from pyspark.sql.types import *

schema = StructType().add("key", LongType()).add("val", PythonOnlyUDT())
df = spark.createDataFrame([(i % 3, PythonOnlyPoint(float(i), float(i))) for i in range(10)], schema=schema)
df.show()

File "/home/spark/python/lib/py4j-0.10.1-src.zip/py4j/protocol.py", line 312, in get_return_value py4j.protocol.Py4JJavaError: An error occurred while calling o36.showString.
: java.lang.RuntimeException: Error while decoding: scala.MatchError: org.apache.spark.sql.types.PythonUserDefinedType@f4ceede8 (of class org.apache.spark.sql.types.PythonUserDefinedType)
...
```
### Second bug: When Python-only UDTs is the element type of ArrayType

```
import pyspark.sql.group
from pyspark.sql.tests import PythonOnlyPoint, PythonOnlyUDT
from pyspark.sql.types import *

schema = StructType().add("key", LongType()).add("val", ArrayType(PythonOnlyUDT()))
df = spark.createDataFrame([(i % 3, [PythonOnlyPoint(float(i), float(i))]) for i in range(10)], schema=schema)
df.show()
```
## How was this patch tested?

PySpark's sql tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @davies

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 1]
```diff
@@ -220,9 +220,15 @@ object RowEncoder {
     CreateExternalRow(fields, schema)
   }
 
-  private def deserializerFor(input: Expression): Expression = input.dataType match {
+  private def deserializerFor(input: Expression): Expression = {
+    deserializerFor(input, input.dataType)
+  }
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: Seems that this method is never used?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: uh? It is the original deserializerFor method and is used below and above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 2]
```diff
@@ -427,8 +427,12 @@ case class MapObjects private(
       case _ => ""
     }
 
+    val inputDataType = inputData.dataType match {
+      case p: PythonUserDefinedType => p.sqlType
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: How about Scala UDT?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: There could be another UDT inside p.sqlType

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala (hunk 1 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: Oh, sorry... Confused by the diff view...

**Comment by vlad17** [comment]
Here's an unresolved example: https://gist.github.com/vlad17/2db8e14972344c693e8a3f03d91c9c8d

**Comment by vlad17** [comment]
Update: looks like the above is just an issue with the __str__ method of udf-returned UDTs, which is a different bug (a bug that's also pretty harmless).

**Comment by vlad17** [comment]
Another update: https://gist.github.com/vlad17/cfcd42f30ea2380df4fb0bfa30dda7ce unresolved

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Scala UDT is already cover by deserializerFor.

**Comment by viirya** [comment]
@vlad17 Thanks! I will look into that issue.

**Comment by mengxr** [comment]
@viirya Do we need to fix this in Spark 2.0? UDTs are private APIs and the only intended use case is Vector/Matrix UDTs for MLlib, which doesn't put vectors or matrices inside an array inside a pipeline. In Spark 2.1, we probably need a formal discussion on merging UDT into Encoder, which could completely change its implementation.

**Comment by viirya** [comment]
@mengxr Although UDTs are private APIs, but as you see from the example, the users can define user classes and corresponding UDTs in Python that will be PythonUserDefinedType. The issues in this PR are not rare cases and it is very possibly PySpark users will hit them during using 2.0.

File: python/pyspark/sql/tests.py [hunk 3]
```diff
@@ -558,6 +558,18 @@ def check_datatype(datatype):
         _verify_type(PythonOnlyPoint(1.0, 2.0), PythonOnlyUDT())
         self.assertRaises(ValueError, lambda: _verify_type([1.0, 2.0], PythonOnlyUDT()))
 
+        schema = StructType().add("key", LongType()).add("val", PythonOnlyUDT())
+        df = self.spark.createDataFrame(
+            [(i % 3, PythonOnlyPoint(float(i), float(i))) for i in range(10)],
+            schema=schema)
+        df.show()
```
Comment (vlad17) [inline python/pyspark/sql/tests.py]: `DataFrame.show()` gives unnecessary stringification, so this test ends up testing unnecessary stuff (in fact it would fail if the UDT didn't have `__str__`. I would use `collect()` to force materialization instead.

File: python/pyspark/sql/tests.py [hunk 4]
```diff
@@ -558,6 +558,18 @@ def check_datatype(datatype):
         _verify_type(PythonOnlyPoint(1.0, 2.0), PythonOnlyUDT())
         self.assertRaises(ValueError, lambda: _verify_type([1.0, 2.0], PythonOnlyUDT()))
 
+        schema = StructType().add("key", LongType()).add("val", PythonOnlyUDT())
```
Comment (vlad17) [inline python/pyspark/sql/tests.py]: This should be in its own test method - it's no longer merely a `test_udt` but rather a `test_simple_udt_in_df`.

File: python/pyspark/sql/tests.py [hunk 5]
```diff
@@ -558,6 +558,18 @@ def check_datatype(datatype):
...
+        df = self.spark.createDataFrame(
+            [(i % 3, PythonOnlyPoint(float(i), float(i))) for i in range(10)],
+            schema=schema)
+        df.show()
+
+        schema = StructType().add("key", LongType()).add("val", ArrayType(PythonOnlyUDT()))
```
Comment (vlad17) [inline python/pyspark/sql/tests.py]: Same here, need its own named unit test (so that it's easier to identify the problem if the test fails) - unit tests should test only one thing, the thing tested here is `test_nested_udt_in_df` (perhaps also worthwhile to check Map works?)

File: python/pyspark/sql/tests.py [hunk 6]
```diff
@@ -558,6 +558,18 @@ def check_datatype(datatype):
...
+        df = self.spark.createDataFrame(
+            [(i % 3, [PythonOnlyPoint(float(i), float(i))]) for i in range(10)],
+            schema=schema)
+        df.show()
+
```
Comment (vlad17) [inline python/pyspark/sql/tests.py]: Missing a unit test for "counterexample 2", nested complex udt struct with mixed types: ArrayType(StructType(simple sql type, udt))

File: python/pyspark/sql/tests.py (hunk 3 above)
Comment (viirya) [inline python/pyspark/sql/tests.py]: This test only fails when using `show()`.

**Comment by viirya** [comment]
ping @vlad17 @davies @liancheng Any thing else?

**Comment by viirya** [comment]
also cc @yhuai @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Why do we need to handle python udf different from scala udf?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: python udf has no userClass. So regular handling of scala udf will be failed.

Comment (vlad17) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: This probably needs a comment explaining the reasoning behind the code. Also, why are we allowed to use p.sqlType for inputDataType serialization in the case of a python UDT? The UDT's deserialize() still has to be called for each element of the array, right?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: > Scala UDT is already cover by deserializerFor.

Can we cover python udt in deserializerFor too?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Yes. This PR adds it.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: when will the `inputData.dataType` be python udt?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: It's really strange we handle python udt here, can we handle it before pass it to `MapObjects`?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: >  when will the inputData.dataType be python udt?

The first bug I showed in the PR description.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: I think it's fixed by https://github.com/apache/spark/pull/13778/files#diff-47e9c0787b1c455e5bd4ad7b65df3436R209 . Can you double check it? If we revert this change, will test fail again?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: ok. let me check it.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: No. The test will be failed.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Is it possible to catch the python udt before passing it to `MapObjects`? I'm kind of worried about leaking python udt to a lot of places, we should handle them just in a few places.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: ok. let me try it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 7]
```diff
@@ -427,8 +430,9 @@ case class MapObjects private(
       case _ => ""
     }
 
+    val inputDT = inputDataType.getOrElse(inputData.dataType)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: @cloud-fan I think there is no way to easily catch python udt here. The approach I use now is to pass a datatype (python udt's sqlType) to MapObjects.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 8]
```diff
@@ -349,11 +349,12 @@ object MapObjects {
   def apply(
       function: Expression => Expression,
       inputData: Expression,
-      elementType: DataType): MapObjects = {
+      elementType: DataType,
+      inputDataType: Option[DataType] = None): MapObjects = {
```
Comment (vlad17) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: I've got quite a few problems with this default:
1. It looks like it's here to avoid coding work in most places, not because it's an obvious value for apply() to take on
2. The additional parameter is completely undocumented and call sites have no mention of it.

Also, the use of option in the case class constructor is a bit obtuse.

Here is my suggestion:
1. Remove the default.
2. If the option in apply() is None, then pass inputDataType.getOrElse(inputData.dataType) as the inputDataType : DataType to the case class constructor, which uses the parameter data type without any hidden logic (as it does now).
3. Document the fact that supplying None in apply() triggers this kind of inference.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: At most scenarios, we don't need to specify inputDataType. So I think it is indeed the obvious value. I agreed that the use of option here is not good. I will change it and have it documented.

**Comment by viirya** [comment]
@cloud-fan @vlad17 Is this change good for your now? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 9]
```diff
@@ -346,14 +346,38 @@ case class LambdaVariable(value: String, isNull: String, dataType: DataType) ext
...
+  }
+
+  /**
+   * Construct an instance of MapObjects case class.
+   *
+   * @param function The function applied on the collection elements.
+   * @param inputData An expression that when evaluated returns a collection object.
+   * @param elementType The data type of elements in the collection.
+   * @param inputDataType The explicitly given data type of inputData to override the
+   *                      data type inferred from inputData (i.e., inputData.dataType).
```
Comment (vlad17) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Would you mind adding in the documentation why this would ever be necessary (i.e., in the array of python UDT case, what's wrong with inputData.dataType)?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: OK. Added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 10]
```diff
@@ -346,14 +346,47 @@ case class LambdaVariable(value: String, isNull: String, dataType: DataType) ext
...
+   */
   def apply(
       function: Expression => Expression,
       inputData: Expression,
       elementType: DataType): MapObjects = {
     val loopValue = "MapObjects_loopValue" + curId.getAndIncrement()
     val loopIsNull = "MapObjects_loopIsNull" + curId.getAndIncrement()
     val loopVar = LambdaVariable(loopValue, loopIsNull, elementType)
-    MapObjects(loopValue, loopIsNull, elementType, function(loopVar), inputData)
+    MapObjects(loopValue, loopIsNull, elementType, function(loopVar), inputData, None)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: It is possibly that inputData is unresolved yet. We can't just pass in the data type of inputData. So I still make inputDataType as Option[DataType] below.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 11]
```diff
@@ -374,13 +407,15 @@ object MapObjects {
  * @param lambdaFunction A function that take the `loopVar` as input, and used as lambda function
  *                       to handle collection elements.
  * @param inputData An expression that when evaluated returns a collection object.
+ * @param inputDataType The dataType of inputData.
```
Comment (vlad17) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Document that it's an optional and say default behavior is to use the resolved .dataType of inputData by default.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: OK.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Added.

**Comment by viirya** [comment]
ping @cloud-fan @vlad17 Any thing else?

**Comment by vlad17** [comment]
LGTM +1

**Comment by cloud-fan** [comment]
From another point of view, is it necessary to propagate the python UDF from python side to jvm side? IIUC the serialization of python UDT happens at python side, and the jvm side can only see binary for python data, there is nothing we can do at java side. Correct me if I am wrong, thanks.

**Comment by viirya** [comment]
Python UDT in python side only serializes the python data to sql type defined in the Python UDT. The problem now is happened at the serialization to row in java side on the serialized python data. I think it can not be certain that the serialized python data doesn't need the serialization in java side.

**Comment by viirya** [comment]
ping @cloud-fan any more concern?

**Comment by cloud-fan** [comment]
Can you point out where we catch `PythonUserDefinedType` and do special serialization at java side? It looks to me that we just get its corresponding sql type.

**Comment by viirya** [comment]
Oh, I mean they should be serialized/deserialized by pickler.

**Comment by cloud-fan** [comment]
yea, so whatever the data type is(python udt or normal sql type), at java side there is no difference, the data is converted to corrected format by pickler. That's why I think maybe it's possible to just pass the corresponding sql type of python udt to java side.

My only concern is, sometimes we use the schema of java dataframe as the schema at python side. If we don't pass python udt to java side, the udt information will be lost. @viirya do you mind give it a try? thanks!

**Comment by viirya** [comment]
@cloud-fan I just checked the python UDT. In python side, we will serialize the python UDT to binary. The python UDT passed to java includes the binary. Then in python side, in the worker we will deserialize the binary back to python UDT and use it for sql data serialization.

Because that, I think we can't just pass the sql type of python UDT to java side. What you think?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 12]
```diff
@@ -346,14 +346,47 @@ case class LambdaVariable(value: String, isNull: String, dataType: DataType) ext
...
+   * @param function The function applied on the collection elements.
+   * @param inputData An expression that when evaluated returns a collection object.
+   * @param elementType The data type of elements in the collection.
+   * @param inputDataType The explicitly given data type of inputData to override the
+   *                      data type inferred from inputData (i.e., inputData.dataType).
+   *                      When Python UDT whose sqlType is an array, the deserializer
+   *                      expression will apply MapObjects on it. However, as the data type
+   *                      of inputData is Python UDT, which is not an expected array type
+   *                      in MapObjects. In this case, we need to explicitly use
+   *                      Python UDT's sqlType as data type.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: As we have to mention python udt in `MapObjects` anyway, I think it makes more sense to add the python udt handling in `MapOjects` directly.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Do you mean the early commit? I remember it is the first approach I take.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: But I think it exposes python udt to `MapObjects`?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: But now we expose too. Readers have to know about python udt to understand this code.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Hmm, ok. Let me update it.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: If we can hide python udt from `MapObjects` entirely, it worth to do. But looks like we can't, and I think then it makes more sense to expose python udt more explicitly.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Making sense. I will update it later.

**Comment by viirya** [comment]
@cloud-fan Updated. Please take a look. Thanks.

**Comment by viirya** [comment]
ping @cloud-fan Please see if this is ok for you now. Thanks.

**Comment by viirya** [comment]
ping @cloud-fan Can you review this? Thanks.

**Comment by viirya** [comment]
ping @liancheng @yhuai Maybe you can review this too?

**Comment by viirya** [comment]
ping @cloud-fan Can you check if this is good for you now? It is for a while. Thanks.

**Comment by viirya** [comment]
ping @cloud-fan What do you think about this? Can we merge it now? Thanks.

**Comment by viirya** [comment]
ping @cloud-fan Do you miss this? Or you have other concern? Please let me know. Thanks.

**Comment by viirya** [comment]
ping @cloud-fan again, this is waiting for a while. Do you have time to look at again? Thanks.

**Comment by davies** [comment]
LGTM, merging this into master and 2.0 branch, thanks!
